navigation/scripts/movement.py

```diff
@@ -0,0 +1,193 @@
+
+import rospy
+import math
+import numpy as np 
+import matplotlib.pyplot as plt 
+from nav_msgs.msg import Odometry, Path
+from geometry_msgs.msg import Twist, Pose, Point32
+from shapely.geometry import Point, LineString, Polygon
+from collections import namedtuple
+from navigation.msg import PolyArray, PointArray
+from navigation.srv import Planner, PlannerRequest, PlannerResponse
+import tf
+import utils
+import collections
+import shapely
+import json
+
+
+Orientation = namedtuple('Orientation', ['roll', 'pitch', 'yaw'])
+DISTMIN = 0.01
+
+class Bot():
+
+    def __init__(self):
+        self.final_goal = Point()
+        self.path = []
+        self.path_use = []
+        self.vel_pub = rospy.Publisher('/cmd_vel', Twist, queue_size=10)
+        self.odom_sub = rospy.Subscriber('/odom', Odometry, self.odom_update)
+        #self.path_sub = rospy.Subscriber('/path', Path, self._update_path)
+        self.obstacle_sub = rospy.Subscriber('obstacles', PolyArray, self.obstacle_update)
+        self.obstacles = PolyArray()
+        self.obstacle_list = []
+        self.path_planner = rospy.ServiceProxy('rrt_planner_service', Planner)
+        self.position = Point(0, 0, 0)
+        self.orientation = Orientation(0, 0, 0)
+        self.velocity = Twist()
+        self.move_to_goal = False
+        self.goal = Point()
+        self.goal_reached = False
+        self.rate = rospy.Rate(10)
+        self.position0 = Point(0, 0, 0)
+        self.orientation0 = Orientation(0, 0, 0)
+        self.position_transformed = Point(0, 0, 0)
+        
+    def odom_update(self, msg):
+        ## callback for odom sub
+        self.position = msg.pose.pose.position
+        self.orientation = Orientation(*tf.transformations.euler_from_quaternion([
+            msg.pose.pose.orientation.x,
+            msg.pose.pose.orientation.y,
+            msg.pose.pose.orientation.z,
+            msg.pose.pose.orientation.w
+        ]))
+        self.position_transformed = Point(self.position.x - self.position0.x, self.position.y - self.position0.y, 0)
+        if len(self.path) < 1 and not self.goal_reached:#or not self.check_intersection(self.path_points, [[(point.x, point.y) for point in polygon.points] for polygon in self.obstacles.polygons]):
+            print('Service call')
+            self.get_path()
+
+        if len(self.path) >= 1:
+            print('into path_use func')
+            try:
+                self.path_use = [Point(p[0], p[1]) for p in utils.transform(
+                        LineString([(p.x, p.y) for p in self.path]), self.position_transformed, self.orientation0).coords]
+            except ValueError:
+                self.path_use = [Point(p[0], p[1]) for p in utils.transform(
+                        Point([(p.x, p.y) for p in self.path]), self.position_transformed, self.orientation0).coords]
+
+        #print([(p.x, p.y) for p in self.path_use][0])        
+        self._set_goal()
+        #self.vel_update()
+        self.rate.sleep()
```

Comment:
```suggestion
```

Reply:
okay.. btw what's the issue with "rate.sleep()"? It's only trying to maintain the loop at a certain frequency

Reply:
Its put in a subscriber callback which is being called with the frq of the topic. If the rate of the topic is faster than the rate set in the callback there can be issues